<?xml version="1.0"?>
<robot name="px_robot">

  <link name="world"></link>

  <link name="base_link">
    <visual name="">
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://px_test/meshes/Base.STL" scale="1 1 1"/>
      </geometry>
      <material name="dark gray">
        <color rgba="0.1 0.1 0.1 1.0"/>
      </material>
    </visual>
    <!-- <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="0.0 0.0 0.0"/>
      </geometry>
    </collision> -->
    <inertial>
      <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
      <mass value="0.2"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial>
  </link>

  <joint name="world_fixed" type="fixed">
    <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
    <parent link="world"/>
    <child link="base_link"/>
  </joint>

  <link name="link1">
    <visual name="">
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://px_test/meshes/Shoulder.STL" scale="1 1 1"/>
      </geometry>
      <material name="dark gray">
        <color rgba="0.1 0.1 0.1 1.0"/>
      </material>
    </visual>
    <!-- <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="0.0 0.0 0.0"/>
      </geometry>
    </collision> -->
    <inertial>  
      <origin xyz="0.0 0.0 0.005" rpy="0.0 0.0 0.0"/>
      <mass value="0.01"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial> 
  </link>

  <joint name="joint_1" type="revolute">
    <origin xyz="0.0 0.0 0.09" rpy="0.0 0.0 0.0"/>
    <parent link="base_link"/>
    <child link="link1"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.32" upper="2.32" effort="10.0" velocity="1.57"/>
  </joint>

  <transmission name="trans_joint_1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_1">
      <hardwareInterface>hardware_interface/PositionJointInterface </hardwareInterface > 
    </joint >
    <actuator name="motor_joint_1">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator> 
  </transmission>

  <link name="link2">
    <visual name="">
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://px_test/meshes/Bicep.STL" scale="1 1 1"/>
      </geometry>
      <material name="dark gray">
        <color rgba="0.1 0.1 0.1 1.0"/>
      </material>
    </visual>
    <!-- <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="0.0 0.0 0.0"/>
      </geometry>
    </collision> -->
    <inertial>
      <origin xyz="0.0 0.0 0.1" rpy="0.0 0.0 0.0"/>
      <mass value="0.01"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial> 
  </link>

  <joint name="joint_2" type="revolute">
    <origin xyz="0.0 0.0 0.042" rpy="1.5707 0.0 0.0"/>
    <parent link="link1"/>
    <child link="link2"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.32" upper="2.32" effort="10.0" velocity="1.57"/>
  </joint>

  <transmission name="trans_joint_2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_2">
      <hardwareInterface>hardware_interface/PositionJointInterface </hardwareInterface > 
    </joint >
    <actuator name="motor_joint_2">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator> 
  </transmission>

  <link name="link3">
    <visual name="">
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://px_test/meshes/Bicep.STL" scale="1 1 1"/>
      </geometry>
      <material name="dark gray">
        <color rgba="0.1 0.1 0.1 1.0"/>
      </material>
    </visual>
    <!-- <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="0.0 0.0 0.0"/>
      </geometry>
    </collision> -->
    <inertial>
      <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
      <mass value="0.01"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial> 
  </link>

  <joint name="joint_3" type="revolute">
    <origin xyz="0.0 0.1065 0" rpy="0.0 0.0 0.0"/>
    <parent link="link2"/>
    <child link="link3"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.32" upper="2.32" effort="10.0" velocity="1.57"/>
  </joint>

  <transmission name="trans_joint_3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_3">
      <hardwareInterface>hardware_interface/PositionJointInterface </hardwareInterface > 
    </joint >
    <actuator name="motor_joint_3">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator> 
  </transmission>

  <link name="link4">
    <visual name="">
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <mesh filename="package://px_test/meshes/Wrist.STL" scale="1 1 1"/>
      </geometry>
      <material name="dark gray">
        <color rgba="0.1 0.1 0.1 1.0"/>
      </material>
    </visual>
    <!-- <collision>
      <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
      <geometry>
        <box size="0.0 0.0 0.0"/>
      </geometry>
    </collision> -->
    <inertial>
      <origin xyz="0.0 0.0 0.01" rpy="0.0 0.0 0.0"/>
      <mass value="0.01"/>
      <inertia ixx="0.001" ixy="0.0" ixz="0.0" iyy="0.001" iyz="0.0" izz="0.001"/>
    </inertial> -->
  </link>

  <joint name="joint_4" type="revolute">
    <origin xyz="0.0 0.1065 0" rpy="0.0 0.0 0.0"/>
    <parent link="link3"/>
    <child link="link4"/>
    <axis xyz="0 0 1"/>
    <limit lower="-2.32" upper="2.32" effort="10.0" velocity="1.57"/>
  </joint>

  <transmission name="trans_joint_4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="joint_4">
      <hardwareInterface>hardware_interface/PositionJointInterface </hardwareInterface > 
    </joint >
    <actuator name="motor_joint_4">
      <mechanicalReduction>1</mechanicalReduction>
    </actuator> 
  </transmission>

  <link name="tcp_link"></link>

  <joint name="tcp" type="fixed">
    <origin xyz="0.0 0.092 0" rpy="-1.5707 0.0 0.0"/>
    <parent link="link4"/>
    <child link="tcp_link"/>
  </joint>

  <!-- Gazebo plugin for ROS Control -->
  <gazebo>
    <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
      <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType >
    </plugin>
  </gazebo>

</robot>